subsys/bluetooth/controller/ll_sw/nordic/lll/lll.c

```diff
@@ -1005,14 +990,52 @@ static void preempt(void *param)
 
 	/* Preemptor not in pipeline */
 	if (next->prepare_param.param != param) {
+		struct lll_event *next_next = NULL;
+		struct lll_event *e;
 		uint32_t ret;
 
-		/* Start the preempt timeout */
-		ret = preempt_ticker_start(next, NULL, next);
-		LL_ASSERT((ret == TICKER_STATUS_SUCCESS) ||
-			  (ret == TICKER_STATUS_BUSY));
+		/* Find if a short prepare request in the pipeline */
+		do {
+			e = ull_prepare_dequeue_iter(&idx);
+			if (!next_next && e && !e->is_aborted &&
+			    !e->is_resume) {
+				next_next = e;
+			}
+		} while (e && (e->is_aborted || e->is_resume ||
+			       (e->prepare_param.param != param)));
 
-		return;
+		/* No short prepare request in pipeline */
+		if (!e) {
+			/* Start the preempt timeout for next event */
+			ret = preempt_ticker_start(next, NULL, next);
+			LL_ASSERT((ret == TICKER_STATUS_SUCCESS) ||
+				  (ret == TICKER_STATUS_BUSY));
+
+			return;
+		}
+
+		/* FIXME: Abort all events in pipeline before the short
+		 *        prepare event. For now, lets assert when many
+		 *        enqueued prepares need aborting.
+		 */
+		LL_ASSERT(next_next == e);
+
+		/* Abort the prepare that is present before the short prepare */
+		next->is_aborted = 1;
+		next->abort_cb(&next->prepare_param, next->prepare_param.param);
+
+		/* As the prepare queue has been refreshed due to the call of
+		 * abort_cb which invokes the lll_done, find the latest prepare
+		 */
+		idx = UINT8_MAX;
+		do {
+			next = ull_prepare_dequeue_iter(&idx);
+		} while (next && (next->is_aborted || next->is_resume));
```

Comment:
This seems to existing many places in the controller code. 

It seems to dequeue until `next` becomes NULL (queue empty?) or as long as the item is either aborted or `is_resume`. 

It appears to me that this could (should?) be put into a function something like `ull_prepare_dequeue_all_aborted` or something like that